File: output_20260320_065752/qwen3/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

/*@ predicate person(struct person* p; struct person* spouse) =
  p != 0 &*&
  malloc_block_person(p) &*&
  p->spouse |-> spouse;
@*/

struct person {
  char* name;
  struct person* spouse;
};

//@ requires true;
//@ ensures person(result, 0);
struct person *create_person()
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close person(p, 0);
  return p;
}

//@ requires person(this, _) &*& person(other, _);
//@ ensures person(this, other) &*& person(other, this);
void marry(struct person *this, struct person *other)
{
  //@ open person(this, _);
  //@ open person(other, _);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this, other);
  //@ close person(other, this);
}

//@ requires person(this, ?s);
//@ ensures person(this, s) &*& result == s;
struct person* person_get_spouse(struct person* this)
{
  //@ open person(this, ?s);
  struct person* sp = this->spouse;
  //@ close person(this, s);
  return sp;
}

//@ requires person(this, ?s) &*& s != 0 &*& person(s, this);
//@ ensures person(this, 0) &*& person(s, 0);
void divorce(struct person* this)
{
  //@ open person(this, ?s);
  //@ open person(s, this);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close person(this, 0);
  //@ close person(s, 0);
}

//@ requires person(this, ?s) &*& (s != 0 ? person(s, this) : true);
//@ ensures true;
void die(struct person *this)
{
  //@ open person(this, ?s);
  if(this->spouse != 0) {
    //@ open person(s, this);
    this->spouse->spouse = 0;
    //@ close person(s, 0);
  }
  free(this); 
}

//@ requires true;
//@ ensures true;
int main() 
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260319_145930/qwen3/java/Linklist/Spouse.java
/*@ 
predicate Person_spouse(Person p; Person spouse) =
    p.spouse |-> spouse;
@*/

class Person {

    protected Person spouse;

    //@ requires true;
    //@ ensures true;
    public void spouse_symm()
    {
    }

    //@ requires true;
    //@ ensures Person_spouse(this, null);
    public Person()
    {
        //@ close Person_spouse(this, null);
    }
    
    //@ requires Person_spouse(this, ?s);
    //@ ensures Person_spouse(this, s) &*& result == s;
    public Person getSpouse()
    {
        return spouse;
    }
    
    //@ requires Person_spouse(this, _) &*& Person_spouse(other, _);
    //@ ensures Person_spouse(this, other) &*& Person_spouse(other, this);
    protected void setSpouse(Person other)
    {
        //@ open Person_spouse(this, _);
        //@ open Person_spouse(other, _);
        spouse = other;
        other.spouse = this;
        //@ close Person_spouse(this, other);
        //@ close Person_spouse(other, this);
    }
    
    //@ requires Person_spouse(this, ?s) &*& s != null &*& Person_spouse(s, this);
    //@ ensures Person_spouse(this, null) &*& Person_spouse(s, null);
    protected void clearSpouse()
    {
        //@ open Person_spouse(this, s);
        //@ open Person_spouse(s, this);
        spouse.spouse = null;
        spouse = null;
        //@ close Person_spouse(this, null);
        //@ close Person_spouse(s, null);
    }
    
    //@ requires Person_spouse(this, _) &*& Person_spouse(other, _);
    //@ ensures Person_spouse(this, other) &*& Person_spouse(other, this);
    void marry(Person other)
    {
        other.setSpouse(this);
    }
    
    //@ requires Person_spouse(this, ?s) &*& s != null &*& Person_spouse(s, this);
    //@ ensures Person_spouse(this, null) &*& Person_spouse(s, null);
    void divorce()
    {
        spouse.clearSpouse();
    }

}

class Program {

    //@ requires true;
    //@ ensures true;
    public static void main(String[] args)
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}